File: python/Prog_석유시추.py
from collections import deque

dr = [-1,0,1,0]
dc = [0,-1,0,1]

R = 0
C = 0
land = [[]]


def bfs(r,c, visited, oil_cnt) :
    q = deque([[r,c]])
    visited[r][c] = True
    cnt = 1 
    
    c_set = set()
    
    while q :
        cr,cc = q.popleft()
        c_set.add(cc)
        for i in range(4) :
            nr = cr + dr[i]
            nc = cc + dc[i]
            if 0<=nr<R and 0<=nc<C and not visited[nr][nc] and land[nr][nc] == 1:
                visited[nr][nc] = True
                q.append([nr,nc])
                cnt += 1
                
    for c in c_set :
        oil_cnt[c] += cnt
    

def solution(land_):
    global R,C,land
    land = land_
    R = len(land)
    C = len(land[0])
    
    oil_cnt = [0 for _ in range(C)]
    
    visited = [[False for _ in range(C)] for _ in range(R)] 
    
    for r in range(R) :
        for c in range(C) :
            if not visited[r][c] and land[r][c] == 1: 
                bfs(r,c, visited, oil_cnt)
    
    return max(oil_cnt)

File: java/src/Prog_석유시추.java
import java.util.*;

class Solution {
    public static class Point {
        public int r;
        public int c;
        
        Point(int r, int c) {
            this.r = r;
            this.c = c;
        }
    }
    
    public int[] dr = {-1,0,1,0};
    public int[] dc = {0,-1,0,1};
    public int R = 0;
    public int C = 0;
    
    public boolean visited[][] = new boolean[500][500];
    public boolean visited_oil[][] = new boolean[500][500];
    
    public int getCnt(int r, int c, int[][] land) {
        visited[r][c] = true;
        Queue<Point> q = new LinkedList<>();
        q.add(new Point(r,c));
        
        int ret = 0;
        
        while (!q.isEmpty()) {
            Point cur = q.poll();
            ret++;
            
            for (int i=0;i<4;i++) {
                int nr = cur.r + dr[i];
                int nc = cur.c + dc[i];
                
                if (nr>=0 && nr<R && nc>=0 && nc<C && land[nr][nc] == 1 && !visited[nr][nc]) {
                    visited[nr][nc] = true;
                    q.add(new Point(nr,nc));
                }
            }
        }
        
        return ret;
    }
    
    public void bfs(int r, int c, int[][] land, int[] oil, int cnt) {
        boolean[] visited_c = new boolean[500];
        
        Queue<Point> q = new LinkedList<>();
        q.add(new Point(r,c));
        visited_oil[r][c] = true;
        
        while(!q.isEmpty()) {
            Point cur = q.poll();
            
            if (!visited_c[cur.c]) {
                visited_c[cur.c] = true;
                oil[cur.c] += cnt;
            }
            
            for (int i=0;i<4;i++) {
                int nr = cur.r + dr[i];
                int nc = cur.c + dc[i];
                
                if (nr>=0 && nr<R && nc>=0 && nc<C && !visited_oil[nr][nc] && land[nr][nc] == 1) {
                    visited_oil[nr][nc] = true;
                    q.add(new Point(nr,nc));
                }
            }
        }
    }
    
    public int solution(int[][] land) {
        R = land.length;
        C = land[0].length;
        
        int[] oil = new int[500];
        for (int r=0;r<R;r++) {
            for (int c=0;c<C;c++) {
                if (land[r][c] == 1 && !visited[r][c]) {
                    int cnt = getCnt(r,c, land);
                    bfs(r,c, land, oil, cnt);
                }
            }
        }
        
        int ans = 0;
        for (int i=0;i<C;i++) {
            ans = Math.max(ans, oil[i]);
        }
        return ans;
    }
}